java/src/jmri/jmrit/operations/rollingstock/RollingStock.java

```diff
@@ -813,7 +815,7 @@ public void propertyChange(java.beans.PropertyChangeEvent e) {
                         if (e.getNewValue() != null) {
                             Location newLocation = locationManager.getLocationByReporter((jmri.Reporter) e.getNewValue());
                             if (newLocation != getLocation())
-                                setLocation(newLocation, null);
+                                setWhereLastSeen(newLocation);
```

Comment:
Sent from my iPad

On Aug 10, 2016, at 5:20 PM, Daniel Boudreau <notifications@github.com> wrote:

> Hi Paul, Sorry if I messed up your scripts. The issue is that I have a user that wants to comparet
> 
> what the operations program believes the car is, versus what is actually provided by readers.
> 
> Previous to my latest change, when a car went through a reader, the code would update the
> 
> car’s location, and null out the car’s track. Then if the operations program detected train
> 
> movement, would override the reader location, and update the car’s track.
> 

Correct, and that's the correct behavior when a reader is located on the mainline or an 
arrival/departure track in a yard.  

> The user asked that the operations car’s xml file provide him with both the location that operation
> 
> believes the car is and what the reporters registered. He wants to write his own program to find
> 
> those discrepancies using the operations xml car file. I was planning on adding more code that
> 
> would highlight all cars in the table where a car’s operation location doesn’t match the tag’s last
> 
> known location. I’m also planning on logging all the “where last seen” location changes so a user
> 
> can review how the car got to a particular location.
> 
If you only have readers at key points ( entering and exiting yards and at junction points ) 


>> 1) Why are you duplicating the When Last Seen and Where Last Seen properties of the ID Tag?
>> 
>> Those should be persistent values associated with the ID tag, so we can look those up, and
>> 
>> shouldn't need to be stored with the rolling stock.
>> 
> My bad. Didn’t realize that the tag was keeping those two things. I see that now. The user also
> 
> wants to have a reader for each track in staging so he can have a wheel report by track. So I’ll
> 
> either need to save this info in operations, or add it to the ID tag.
> 
That's actually overkill.

You can build a wheel report by track with a single reader at the yard thought using switch 
position information to determine 
>> 2) The code that sets Where Last Seen and When Last Seen instead of the Location and the
>> 
>> Date also need to store the Location and the Date. We may not have ID tag readers at ever
>> 
>> location, but when a car passes a reader, that reader is associated with a location, and we
>> 
>> know the car MUST be at that location.
>> 
> I agree. But the two fields were being updated by two very different mechanisms.
> 
No.  The location and date were both updated when the reader sees a tag and notified the 
reader that either the last.  Yes, the location also changes when the train location is updated 
by the user, but that is after the point at which the information from the readers is useful.

The readers tell us that a piece of rolling stock passed a location at a given time.  By the time
the user moves the train to a new location, the expectation is that the car order will have 
changed, so we can't rely on the old tag generated location and date to be correct.
>> The 2nd part of this is VERY important to me, because it changes the previous behaviour.
>> 
>> This code change breaks scripts that I've got locally to print wheel reports.
>> 
> Are you using operations and readers at the same time? I don’t understand how this would work.
> 
> I would think if the ID tag has the car’s last known location and date, that your scripts could pull
> 
> this from the car’s ID tag and not from the operation’s database.
> 
Yes, I am using operations with readers. Remember, I'm the one that put The code in place to 
update the locations to begin with.

I am actually replicating the information that RFID readers provide the prototype, which is that
a piece of rolling stock passed a reader at a key point location.  When the train moves ( via 
either the move or the terminate button ) that represents another operation on the prototype ( which
 is a clerk making a notation in the computer that a train has moved ).

More importantly though is that you changed the behavior of something that has been in 
production and at least a few people have come to use.  We have generally been trying to avoid 
changing the expected behavior of the API at a moments notice, and this change definitely does 
that.

Incidentally, one of the projects I have planned is to also be able to associate an ID tag with a 
train.  The ID tags will be functionally replace OSing the train location, and will be used to move
the train along the route.

For my own purposes, the ID tag will be credit card sized RFID swiped on a fascia mounted 
reader, but there will be options to use the tag from a piece of rolling stock in the train ( cabooses,
locomotives,etc ) as well.

The other thing project that goes along with the fascia mounted RFID readers to update train 
positions is a fascia mounted receipt printer that will print out the wheel reports and/or switchlists 
at key locations. I'm in the process of installing one of these at a new staging yard I started 
building recently.  

Paul